### PR TITLE
Add in TTL work, with tests

### LINKED ISSUE #1414: Enhancement request to include TTL for Documents in Astra DB
A client currently using Couchbase has indicated Astra DB is missing the Document Expiry feature. We like to suggest an enhancement request to include TTL for Documents in Astra DB.

Please refer to Couchbase documentation for details: https://docs.couchbase.com/server/current/eventing/eventing-examples-docexpiry.html

### PR DESCRIPTION
**Pull request by EricBorczuk**
**What this PR does**:
Allows assigning a TTL to a document. Updated version of https://github.com/stargate/stargate/pull/1428

**Which issue(s) this PR fixes**:
Fixes #1414 

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [x] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java [hunk 1]
```diff
@@ -110,6 +110,39 @@ public ReactiveDocumentService(
...
+    try {
+      Integer ttlValue = Integer.parseInt(ttlRaw);
+      if (subPath.isEmpty()) {
+        return ttlValue;
+      } else {
+        throw new ErrorCodeRuntimeException(ErrorCode.DOCS_API_INVALID_TTL_NUMBER);
+      }
+    } catch (NumberFormatException e) {
+      if (!subPath.isEmpty()) {
+        if (!ttlRaw.equalsIgnoreCase("auto")) {
```
Comment (tatu-at-datastax) [inline restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:134]: I would suggest explicitly looking for "auto" first since exceptions are rather expensive things to throw -- so creating and catching `NumberFormatException`s can be wasteful. More efficient to check for special case(s) first.
Same goes for empty String actually, probably better to check those 2 before attempting number parsing.

File: restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentDB.java [hunk 2]
```diff
@@ -120,6 +123,32 @@ public Keyspace getKeyspace(String keyspaceName) {
...
+          .map(row -> row.getInt(String.format("ttl(%s)", DocsApiConstants.LEAF_COLUMN_NAME)))
+          .filter(Objects::nonNull)
+          .mapToInt(Integer::valueOf)
+          .max()
+          .getAsInt();
+    } catch (NoSuchElementException e) {
+      return null;
+    }
+  }
+
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentDB.java:151]: We should avoid adding any new things in the `DocumentDB` and rather as such functionality to the `DocumentSearchService`.. I am also fine to have a separate `DocumentTTLService`, but point is that this is implemented in the same way as we do it for the document read:

* dedicated query builder
* tests for query builder
* reactive service implementation that does separate prepare and execute, and returns a Single<Integer> (can be empty in case there is no ttl)
* test for that service implementation

File: restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentDB.java [hunk 3]
```diff
@@ -288,25 +317,6 @@ public void tableExists(String keyspaceName, String tableName) {
...
-        .prepare(
-            dataStore
-                .queryBuilder()
-                .insertInto(info.getKeyspace(), info.getTable())
-                .value(markers)
-                .timestamp()
-                .build())
-        .join();
-  }
-
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentDB.java:309]: good catch! This is exactly what I am talking about, removing peace by peace from the `DocumentDB`.. For V2 we need to do this..

File: restapi/src/main/java/io/stargate/web/docsapi/resources/ReactiveDocumentResourceV2.java [hunk 4]
```diff
@@ -391,6 +424,12 @@ public void patchDoc(
           @NotNull(message = "payload not provided")
           @NotBlank(message = "payload must not be empty")
           String payload,
+      @ApiParam(
+              value = "Include this to put a time-to-live on the document in the data store",
+              required = false)
+          @QueryParam("ttl")
+          @Min(value = 1, message = "TTL value must be a positive integer.")
+          Integer ttl,
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/resources/ReactiveDocumentResourceV2.java:432]: wait this is patch, you can not submit the TTL here? we need a different parameter like `ttl-auto=true`, right?

File: restapi/src/main/java/io/stargate/web/docsapi/resources/ReactiveDocumentResourceV2.java [hunk 5]
```diff
@@ -319,6 +346,12 @@ public void putDocPath(
           @NotNull(message = "payload not provided")
           @NotBlank(message = "payload must not be empty")
           String payload,
+      @ApiParam(
+              value =
+                  "Include this to put a time-to-live on the document in the data store; use 'auto' on a sub-document to make the TTL match that of the parent document.",
+              required = false)
+          @QueryParam("ttl")
+          String ttl,
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/resources/ReactiveDocumentResourceV2.java:354]: let's have a different name for the param and make it a boolean, see other comment

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java [hunk 6]
```diff
@@ -125,6 +158,7 @@ public ReactiveDocumentService(
       String namespace,
       String collection,
       String payload,
+      Integer ttl,
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:161]: add java doc

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java [hunk 7]
```diff
@@ -305,9 +344,11 @@ public ReactiveDocumentService(
       String collection,
       String documentId,
       String payload,
-      ExecutionContext context) {
+      String ttl,
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:347]: please no string! Here integer is fine, this is where we allow the ttl to be specified..

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java [hunk 8]
```diff
@@ -330,7 +372,10 @@ public ReactiveDocumentService(
       String documentId,
       List<String> subPath,
       String payload,
-      ExecutionContext context) {
+      String ttl,
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:375]: same here.. as we need to handle both update of the whole document with ttl int and the ttl auto I propse:

1. `updateDocument` accepts ttl int and calls `updateDocumentInternal`
2. add `updateSubDocument` with ttl-auto boolean and if it's `true` resolves the ttl and calls the `updateDocumentInternal` with the value.
3. this method is renamed to `updateDocumentInternal` and accepts only ttl as int

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java [hunk 9]
```diff
@@ -381,9 +428,11 @@ public ReactiveDocumentService(
       String collection,
       String documentId,
       String payload,
-      ExecutionContext context) {
+      String ttl,
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:431]: patch only accepts ttl-auto boolean, right?

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java [hunk 10]
```diff
@@ -407,7 +457,10 @@ public ReactiveDocumentService(
       String documentId,
       List<String> subPath,
       String payload,
-      ExecutionContext context) {
+      String ttl,
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:460]: same here

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java [hunk 11]
```diff
@@ -1036,9 +1090,11 @@ public ReactiveDocumentService(
       String keyspace,
       String collection,
       String id,
+      String ttl,
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:1093]: Integer

**Review by ivansenic — CHANGES_REQUESTED** [review]
We need more clear service API and execution paths.. Main thing is to go away from the ttl as string and split into two params.. There is a bunch of needed changes here, so take your time.. if you wish we can split this into multiple PRs, if this is easier for you, for example:

1. PR-1 extension of the query builder and the DocumentWriteService to support TTL
2. PR-2 add service to provide current TTL for a document (current DocumentDB impl, but see suggestion)
3. PR-3 rest of the needed changes, HTTP API and reactive document service

This would also make it easier to review, WDYT?

**Comment by EricBorczuk** [comment]
Instead of the separation you suggested, I will be putting the bulk of this step:
```PR-2 add service to provide current TTL for a document (current DocumentDB impl, but see suggestion)```

In a separate PR.

File: restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentDB.java [hunk 12]
```diff
@@ -120,6 +123,32 @@ public Keyspace getKeyspace(String keyspaceName) {
     return dataStore.schema().keyspaces();
   }
 
+  public Integer getTtlForDocument(String namespace, String collection, String docId) {
```
Comment (EricBorczuk) [inline restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentDB.java:126]: This gets removed in https://github.com/stargate/stargate/pull/1684

**Comment by EricBorczuk** [comment]
Of these changes:

PR-1 extension of the query builder and the DocumentWriteService to support TTL
PR-2 add service to provide current TTL for a document (current DocumentDB impl, but see suggestion)
PR-3 rest of the needed changes, HTTP API and reactive document service

#1 and #3 are fixed in this PR! Please review and I'll hold off merging https://github.com/stargate/stargate/pull/1684 to not clutter things

File: restapi/src/main/java/io/stargate/web/docsapi/service/query/search/db/impl/DocumentTtlQueryBuilder.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+/** Simple query builder to get the TTL of a document. */
+public class DocumentTtlQueryBuilder extends AbstractSearchQueryBuilder {
+
+  /** Doc id to target. */
+  private final String documentId;
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/query/search/db/impl/DocumentTtlQueryBuilder.java:40]: this is now not needed, also why Codacy complains..

File: restapi/src/main/java/io/stargate/web/docsapi/resources/ReactiveDocumentResourceV2.java [hunk 14]
```diff
@@ -319,6 +358,11 @@ public void putDocPath(
           @NotNull(message = "payload not provided")
           @NotBlank(message = "payload must not be empty")
           String payload,
+      @ApiParam(
+              value = "Include this to make the TTL match that of the parent document.",
+              required = false)
+          @QueryParam("ttl-auto")
+          Boolean ttlAuto,
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/resources/ReactiveDocumentResourceV2.java:365]: should we also make it a `boolean` here and not the object bool?

File: restapi/src/main/java/io/stargate/web/docsapi/resources/ReactiveDocumentResourceV2.java [hunk 15]
```diff
@@ -319,6 +358,11 @@ public void putDocPath(
           @NotNull(message = "payload not provided")
           @NotBlank(message = "payload must not be empty")
           String payload,
+      @ApiParam(
+              value = "Include this to make the TTL match that of the parent document.",
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/resources/ReactiveDocumentResourceV2.java:362]: can we add that this requires read before write if `true`.. Also to the other methods with `ttlAuto`

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java [hunk 16]
```diff
@@ -305,9 +346,44 @@ public ReactiveDocumentService(
...
+   * @return Document response wrapper containing the generated ID.
+   */
+  public Single<DocumentResponseWrapper<Void>> updateSubDocument(
+      DocumentDB db,
+      String namespace,
+      String collection,
+      String documentId,
+      List<String> subPath,
+      String payload,
+      Boolean ttlAuto,
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:375]: same here should we just have a `boolean`?

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java [hunk 17]
```diff
@@ -381,9 +463,10 @@ public ReactiveDocumentService(
       String collection,
       String documentId,
       String payload,
+      Integer ttl,
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:466]: OK this is a mistake, how can a patch have a ttl as int? should be a `ttlAuto`? This also means you don't have to introduce that new `patchSubDocument` it can still be one method.. Right?

File: restapi/src/test/java/io/stargate/web/docsapi/service/ReactiveDocumentServiceTest.java [hunk 18]
```diff
@@ -197,13 +197,14 @@ public void happyPath() throws Exception {
               eq(collection),
```
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/ReactiveDocumentServiceTest.java:197]: This class is missing 5 tests updates for the TTL:

* write with ttl
* write batch with ttl
* update with ttl
* update sub-path with ttl auto
* patch with ttl auto

**Review by ivansenic — CHANGES_REQUESTED** [review]
Looks good, way better than on the first review.. I left some minor comments for improvements, should be easy to fix.. However, we miss bunch of tests for the reactive document service..

**Comment by ivansenic** [comment]
@EricBorczuk small ping here, not sure if you saw my review from last week.. See no progress here..

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java (hunk 17 above)
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:466]: I believe this method is entirely unused, that's why I missed it. Deleting it in favor of the sub-document impl and renaming that method

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java [hunk 19]
```diff
@@ -442,31 +536,13 @@ public ReactiveDocumentService(
...
-  public Single<Boolean> deleteDocument(
-      DocumentDB db,
-      String namespace,
-      String collection,
-      String documentId,
-      ExecutionContext context) {
-    List<String> subPath = Collections.emptyList();
-    return deleteDocument(db, namespace, collection, documentId, subPath, context);
-  }
-
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:469]: wait why is this deleted now?

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java [hunk 20]
```diff
@@ -1040,15 +1116,19 @@ public ReactiveDocumentService(
...
+                writeService.updateDocument(
+                    db.getQueryExecutor().getDataStore(),
+                    keyspace,
+                    collection,
+                    id,
+                    processedPath,
+                    rows,
+                    ttl,
+                    db.treatBooleansAsNumeric(),
+                    context));
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:1131]: is this additional fix for the push function? :+1:

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java [hunk 21]
```diff
@@ -110,13 +112,46 @@ public ReactiveDocumentService(
...
+  private Single<Integer> determineTtl(
+      DocumentDB db,
+      String namespace,
+      String collection,
+      String documentId,
+      ExecutionContext context) {
+
+    return Single.defer(
+        () -> {
+          authorizeRead(db, namespace, collection);
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:124]: I am starting to think if this is needed? Means if you have only write permissions, you can not actually do update or patch with the TTL auto? Is this correct?

File: restapi/src/main/java/io/stargate/web/docsapi/service/query/DocumentSearchService.java [hunk 22]
```diff
@@ -180,6 +181,16 @@
         .take(1);
   }
 
+  public Flowable<RawDocument> getDocumentTtlInfo(
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/query/DocumentSearchService.java:184]: Java docs missing here.. Not sure if this was added or I missed this..

**Review by ivansenic — APPROVED** [review]
Great stuff.. Have a look on a few remaining comments, especially that authz read on determine TTL..

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java (hunk 19 above)
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:469]: The method is never used...all calls currently are using line 467's `deleteDocument` directly

File: restapi/src/main/java/io/stargate/web/docsapi/service/query/DocumentSearchService.java (hunk 22 above)
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/query/DocumentSearchService.java:184]: Yea this existed in previous reviews, must have just missed javadoc. Will add

File: restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java (hunk 21 above)
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java:124]: Oh that's a good point...Yeah maybe this shouldn't be here then (even though it's technically a read, people should be able to update/patch regardless)
